File: backend/app/routers/runs.py
from __future__ import annotations

import json
from datetime import datetime, timezone
from pathlib import Path
from typing import Any
from uuid import uuid4

from fastapi import APIRouter, Depends, HTTPException
from pydantic import BaseModel, Field

from app.db.session import db_session
from app.runtime.auth import require_admin, require_user
from app.runtime.graph_runner import is_run_active, start_run_async

router = APIRouter(prefix="/workflow-runs", tags=["workflow-runs"])


class StartRunRequest(BaseModel):
    workflow_id: str = Field(min_length=1)
    # Blank falls back to the first approved workspace -- a Telegram message
    # carries no workspace, and the UI already picks one per run.
    workspace_path: str = ""
    graph: dict = {}
    # Values supplied at run time by the workflow's trigger nodes, keyed by the
    # trigger's field name (e.g. {"topic": "How we built repo import"}).
    run_input: dict[str, str] = {}
    trigger_type: str = "manual"


def _public_run(row) -> dict[str, Any]:
    graph_json = row["graph_json"] if "graph_json" in row.keys() else "{}"
    return {
        "id": row["id"],
        "workflow_id": row["workflow_id"],
        "status": row["status"],
        "trigger_type": row["trigger_type"],
        "workspace_path": row["workspace_path"] if "workspace_path" in row.keys() else None,
        "graph": json.loads(graph_json or "{}"),
        # Needed by the Telegram poller to report a result back to the chat.
        "final_report": row["final_report"] if "final_report" in row.keys() else None,
        "created_at": row["created_at"],
        "completed_at": row["completed_at"],
    }


def _public_step(row) -> dict[str, Any]:
    return {
        "id": row["id"],
        "node_id": row["node_id"],
        "node_type": row["node_type"],
        "agent_name": row["agent_name"],
        "agent_role": row["agent_role"],
        "status": row["status"],
        "summary": row["summary"],
        "error": row["error"],
        "started_at": row["started_at"],
        "completed_at": row["completed_at"],
    }


def _public_log(row) -> dict[str, Any]:
    return {
        "id": row["id"],
        "level": row["level"],
        "message": row["message"],
        "metadata": json.loads(row["metadata_json"] or "{}"),
        "created_at": row["created_at"],
    }


def _public_message(row) -> dict[str, Any]:
    return {
        "id": row["id"],
        "agent_run_id": row["agent_run_id"],
        "sender_type": row["sender_type"],
        "sender_name": row["sender_name"],
        "content": row["content"],
        "created_at": row["created_at"],
    }


def _parse_datetime(value: str) -> datetime:
    return datetime.fromisoformat(value.replace("Z", "+00:00"))


def _expire_pending_approvals(run_id: str) -> None:
    now = datetime.now(timezone.utc)
    now_iso = now.isoformat()
    with db_session() as db:
        rows = db.execute(
            """
            SELECT id, workflow_step_run_id, expires_at
            FROM approval_requests
            WHERE workflow_run_id = ? AND status = 'pending' AND expires_at IS NOT NULL
            """,
            (run_id,),
        ).fetchall()
        expired = [row for row in rows if _parse_datetime(row["expires_at"]) <= now]
        if not expired:
            return
        for row in expired:
            db.execute(
                "UPDATE approval_requests SET status = 'expired', resolved_at = ? WHERE id = ? AND status = 'pending'",
                (now_iso, row["id"]),
            )
            if row["workflow_step_run_id"]:
                db.execute(
                    "UPDATE workflow_step_runs SET status = 'cancelled', completed_at = ? WHERE id = ?",
                    (now_iso, row["workflow_step_run_id"]),
                )
                db.execute(
                    "UPDATE agent_runs SET status = 'cancelled', completed_at = ?, error = ? WHERE id = ?",
                    (now_iso, "Approval expired without response.", row["workflow_step_run_id"]),
                )
        db.execute(
            "UPDATE workflow_runs SET status = 'cancelled', completed_at = ? WHERE id = ? AND status = 'waiting_approval'",
            (now_iso, run_id),
        )
        db.execute(
            "INSERT INTO run_logs (id, workflow_run_id, level, message, metadata_json) VALUES (?, ?, 'warn', ?, ?)",
            (str(uuid4()), run_id, "Run cancelled: approval expired without response.", json.dumps({"approval_status": "expired"})),
        )


def _ensure_pending_approval_open(run_id: str, approval_id: str):
    _expire_pending_approvals(run_id)
    with db_session() as db:
        row = db.execute("SELECT * FROM approval_requests WHERE id = ? AND workflow_run_id = ?", (approval_id, run_id)).fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Approval request not found.")
    if row["status"] != "pending":
        raise HTTPException(status_code=400, detail=f"Approval already resolved: {row['status']}")
    return row


def _get_approval(run_id: str, approval_id: str):
    _expire_pending_approvals(run_id)
    with db_session() as db:
        row = db.execute("SELECT * FROM approval_requests WHERE id = ? AND workflow_run_id = ?", (approval_id, run_id)).fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Approval request not found.")
    return row


def _normalize_workspace_path(path: str) -> str:
    return str(Path(path).expanduser().resolve())


def _approved_workspace_path(path: str) -> str:
    requested = Path(_normalize_workspace_path(path))
    with db_session() as db:
        rows = db.execute("SELECT path FROM runtime_workspaces WHERE is_active = 1").fetchall()
    candidates: list[tuple[int, str]] = []
    for row in rows:
        approved = Path(_normalize_workspace_path(row["path"]))
        if requested == approved or approved in requested.parents:
            candidates.append((len(approved.parts), str(approved)))
    if not candidates:
        raise HTTPException(status_code=403, detail="Workspace path is not approved for workflow execution.")
    candidates.sort(key=lambda item: item[0], reverse=True)
    return candidates[0][1]


@router.post("")
def start_run(request: StartRunRequest, user: dict = Depends(require_admin)) -> dict[str, Any]:
    run_id = str(uuid4())
    workspace_path = request.workspace_path.strip()
    if not workspace_path:
        # Fall back to the workflow's OWN workspace, never a global default:
        # running workflow A against workflow B's repo would write to the wrong tree.
        with db_session() as db:
            row = db.execute(
                "SELECT workspace_path FROM workflows WHERE id = ?", (request.workflow_id,)
            ).fetchone()
        workspace_path = (row["workspace_path"] if row else "") or ""
        if not workspace_path:
            raise HTTPException(
                status_code=400,
                detail="This workflow has no workspace set. Open it in the builder, pick a repository, and save.",
            )
    workspace_path = _approved_workspace_path(workspace_path)

    # resolve graph — use saved workflow graph if not provided
    graph = request.graph
    if not graph.get("nodes"):
        with db_session() as db:
            wf = db.execute("SELECT graph_json FROM workflows WHERE id = ?", (request.workflow_id,)).fetchone()
        if not wf:
            raise HTTPException(status_code=404, detail="Workflow not found.")
        graph = json.loads(wf["graph_json"] or "{}")

    with db_session() as db:
        db.execute(
            "INSERT INTO workflow_runs (id, workflow_id, status, trigger_type, graph_json, workspace_path, run_input_json)"
            " VALUES (?, ?, 'queued', ?, ?, ?, ?)",
            (run_id, request.workflow_id, request.trigger_type, json.dumps(graph),
             workspace_path, json.dumps(request.run_input or {})),
        )

    # Remember it so a trigger-started run uses the same repo the UI last ran against.
    with db_session() as db:
        db.execute("UPDATE workflows SET workspace_path = ? WHERE id = ?",
                   (workspace_path, request.workflow_id))

    start_run_async(run_id, request.workflow_id, graph, workspace_path, request.run_input or {})

    return {"run_id": run_id, "status": "queued", "workflow_id": request.workflow_id, "workspace_path": workspace_path}


@router.get("")
def list_runs(
    workflow_id: str | None = None,
    limit: int = 100,
    _: dict = Depends(require_user),
) -> list[dict[str, Any]]:
    """Recent runs, newest first.

    The limit is a caller-visible parameter rather than a hidden 20: callers were
    treating the truncated page as a total, so a dashboard tile froze at 20 once
    the 21st run existed. Anything wanting real totals should use /stats.
    """
    limit = max(1, min(limit, 500))
    with db_session() as db:
        if workflow_id:
            rows = db.execute(
                "SELECT * FROM workflow_runs WHERE workflow_id = ? ORDER BY created_at DESC LIMIT ?",
                (workflow_id, limit),
            ).fetchall()
        else:
            rows = db.execute(
                "SELECT * FROM workflow_runs ORDER BY created_at DESC LIMIT ?", (limit,)
            ).fetchall()
    return [_public_run(r) for r in rows]


@router.get("/stats")
def run_stats(window_hours: int = 24, _: dict = Depends(require_user)) -> dict[str, Any]:
    """Aggregates computed in SQL over the whole table, not a page of it.

    Counting from a truncated list is how the attention banner could report
    "All clear" while failures sat just outside the loaded window.
    """
    window_hours = max(1, min(window_hours, 24 * 30))
    since = f"-{window_hours} hours"
    with db_session() as db:
        totals = db.execute(
            """
            SELECT
              COUNT(*)                                                   AS total,
              SUM(status = 'failed')                                     AS failed,
              SUM(status = 'completed')                                  AS completed,
              SUM(status IN ('running', 'queued', 'waiting_approval'))   AS active
            FROM workflow_runs
            WHERE created_at >= datetime('now', ?)
            """,
            (since,),
        ).fetchone()
        # Active runs are counted without a window: a run started two days ago and
        # still going is exactly what an operator needs to see.
        active_all = db.execute(
            "SELECT COUNT(*) c FROM workflow_runs"
            " WHERE status IN ('running','queued','waiting_approval')"
        ).fetchone()["c"]
        median = db.execute(
            """
            SELECT AVG(secs) AS median FROM (
              SELECT (julianday(completed_at) - julianday(created_at)) * 86400 AS secs
              FROM workflow_runs
              WHERE completed_at IS NOT NULL AND created_at >= datetime('now', ?)
              ORDER BY secs
              LIMIT 2 - (SELECT COUNT(*) FROM workflow_runs
                         WHERE completed_at IS NOT NULL AND created_at >= datetime('now', ?)) % 2
              OFFSET (SELECT (COUNT(*) - 1) / 2 FROM workflow_runs
                      WHERE completed_at IS NOT NULL AND created_at >= datetime('now', ?))
            )
            """,
            (since, since, since),
        ).fetchone()
        # Oldest thing still running. "3 running" says nothing about whether one
        # has been stuck for an hour, which is the question an operator has.
        oldest = db.execute(
            "SELECT MIN(created_at) AS started FROM workflow_runs"
            " WHERE status IN ('running','queued','waiting_approval')"
        ).fetchone()["started"]

        # The same median over the PREVIOUS window, so the tile can show a
        # direction rather than a bare number. A duration with no trend cannot
        # tell you whether things are getting worse.
        prev_since = f"-{window_hours * 2} hours"
        prev_median = db.execute(
            """
            SELECT AVG(secs) AS median FROM (
              SELECT (julianday(completed_at) - julianday(created_at)) * 86400 AS secs
              FROM workflow_runs
              WHERE completed_at IS NOT NULL
                AND created_at >= datetime('now', ?) AND created_at < datetime('now', ?)
              ORDER BY secs
              LIMIT 2 - (SELECT COUNT(*) FROM workflow_runs
                         WHERE completed_at IS NOT NULL
                           AND created_at >= datetime('now', ?) AND created_at < datetime('now', ?)) % 2
              OFFSET (SELECT (COUNT(*) - 1) / 2 FROM workflow_runs
                      WHERE completed_at IS NOT NULL
                        AND created_at >= datetime('now', ?) AND created_at < datetime('now', ?))
            )
            """,
            (prev_since, since, prev_since, since, prev_since, since),
        ).fetchone()

        waiting = db.execute(
            "SELECT COUNT(*) c FROM workflow_runs WHERE status = 'waiting_approval'"
        ).fetchone()["c"]

    current_median = round(median["median"] or 0, 1)
    previous_median = round(prev_median["median"] or 0, 1)
    return {
        "window_hours": window_hours,
        "total": totals["total"] or 0,
        "failed": totals["failed"] or 0,
        "completed": totals["completed"] or 0,
        "active": active_all,
        "waiting_approval": waiting,
        "oldest_active_started_at": oldest,
        "median_duration_seconds": current_median,
        "previous_median_duration_seconds": previous_median,
        # None rather than 0 when there is no prior window: "no change" and "no
        # data to compare" are different claims, and rendering the second as the
        # first invents a trend.
        "median_delta_seconds": (
            round(current_median - previous_median, 1) if previous_median else None
        ),
    }


@router.get("/{run_id}")
def get_run(run_id: str, _: dict = Depends(require_user)) -> dict[str, Any]:
    with db_session() as db:
        row = db.execute("SELECT * FROM workflow_runs WHERE id = ?", (run_id,)).fetchone()
    if not row:
        raise HTTPException(status_code=404, detail="Run not found.")
    return _public_run(row)


@router.get("/{run_id}/steps")
def get_run_steps(run_id: str, _: dict = Depends(require_user)) -> list[dict[str, Any]]:
    with db_session() as db:
        rows = db.execute(
            """
            SELECT ar.*, ws.node_type
            FROM agent_runs ar
            LEFT JOIN workflow_step_runs ws ON ws.id = ar.id
            WHERE ar.workflow_run_id = ?
            ORDER BY ar.started_at ASC
            """,
            (run_id,),
        ).fetchall()
    return [_public_step(r) for r in rows]


@router.get("/{run_id}/logs")
def get_run_logs(run_id: str, _: dict = Depends(require_user)) -> list[dict[str, Any]]:
    with db_session() as db:
        rows = db.execute(
            "SELECT * FROM run_logs WHERE workflow_run_id = ? ORDER BY created_at ASC",
            (run_id,),
        ).fetchall()
    return [_public_log(r) for r in rows]


@router.get("/{run_id}/steps/{step_id}/messages")
def get_step_messages(run_id: str, step_id: str, _: dict = Depends(require_user)) -> list[dict[str, Any]]:
    with db_session() as db:
        rows = db.execute(
            "SELECT * FROM agent_messages WHERE agent_run_id = ? ORDER BY created_at ASC",
            (step_id,),
        ).fetchall()
    return [_public_message(r) for r in rows]


class ApprovalActionRequest(BaseModel):
    note: str = ""


def _resume_payload(run_id: str) -> tuple[str, dict[str, Any], str]:
    with db_session() as db:
        row = db.execute(
            "SELECT workflow_id, graph_json, workspace_path FROM workflow_runs WHERE id = ?",
            (run_id,),
        ).fetchone()
        if not row:
            raise HTTPException(status_code=404, detail="Run not found.")
        workspace_path = row["workspace_path"]
        if not workspace_path:
            workspace = db.execute(
                "SELECT path FROM runtime_workspaces WHERE is_active = 1 ORDER BY updated_at DESC, created_at DESC LIMIT 1"
            ).fetchone()
            if not workspace:
                raise HTTPException(status_code=409, detail="Run cannot resume because no workspace path is stored or approved.")
            workspace_path = workspace["path"]
        return row["workflow_id"], json.loads(row["graph_json"] or "{}"), workspace_path


@router.post("/{run_id}/approve/{approval_id}")
def approve_run(run_id: str, approval_id: str, body: ApprovalActionRequest = ApprovalActionRequest(), _: dict = Depends(require_admin)) -> dict[str, Any]:
    approval = _get_approval(run_id, approval_id)
    if approval["status"] not in ("pending", "approved"):
        raise HTTPException(status_code=400, detail=f"Approval already resolved: {approval['status']}")
    with db_session() as db:
        if approval["status"] == "pending":
            db.execute(
                "UPDATE approval_requests SET status = 'approved', resolved_at = CURRENT_TIMESTAMP, resolution_comment = ? WHERE id = ?",
                (body.note or None, approval_id),
            )
        if approval["workflow_step_run_id"]:
            db.execute(
                "UPDATE workflow_step_runs SET status = 'completed', completed_at = CURRENT_TIMESTAMP WHERE id = ?",
                (approval["workflow_step_run_id"],),
            )
            db.execute(
                "UPDATE agent_runs SET status = 'completed', completed_at = CURRENT_TIMESTAMP, summary = ? WHERE id = ?",
                ("Approved by human reviewer.", approval["workflow_step_run_id"]),
            )
        db.execute(
            "UPDATE workflow_runs SET status = 'running', completed_at = NULL WHERE id = ? AND status = 'waiting_approval'",
            (run_id,),
        )
        db.execute(
            "INSERT INTO run_logs (id, workflow_run_id, level, message, metadata_json) VALUES (?, ?, 'info', ?, ?)",
            (str(uuid4()), run_id, "Approval granted by human reviewer.", json.dumps({"approval_id": approval_id, "approval_status": approval["status"]})),
        )
    resumed = False
    if not is_run_active(run_id):
        workflow_id, graph, workspace_path = _resume_payload(run_id)
        resumed = start_run_async(run_id, workflow_id, graph, workspace_path)
    return {"approved": True, "approval_id": approval_id, "resumed": resumed}


@router.post("/{run_id}/reject/{approval_id}")
def reject_run(run_id: str, approval_id: str, body: ApprovalActionRequest = ApprovalActionRequest(), _: dict = Depends(require_admin)) -> dict[str, Any]:
    approval = _ensure_pending_approval_open(run_id, approval_id)
    with db_session() as db:
        db.execute(
            "UPDATE approval_requests SET status = 'rejected', resolved_at = CURRENT_TIMESTAMP, resolution_comment = ? WHERE id = ?",
            (body.note or None, approval_id),
        )
        if approval["workflow_step_run_id"]:
            db.execute(
                "UPDATE workflow_step_runs SET status = 'failed', completed_at = CURRENT_TIMESTAMP WHERE id = ?",
                (approval["workflow_step_run_id"],),
            )
            db.execute(
                "UPDATE agent_runs SET status = 'failed', completed_at = CURRENT_TIMESTAMP, error = ? WHERE id = ?",
                ("Approval rejected.", approval["workflow_step_run_id"]),
            )
        db.execute(
            "UPDATE workflow_runs SET status = 'failed', completed_at = CURRENT_TIMESTAMP WHERE id = ?",
            (run_id,),
        )
        db.execute(
            "INSERT INTO run_logs (id, workflow_run_id, level, message, metadata_json) VALUES (?, ?, 'warn', ?, ?)",
            (str(uuid4()), run_id, "Run stopped: approval rejected.", json.dumps({"approval_id": approval_id})),
        )
    return {"rejected": True, "approval_id": approval_id}


@router.post("/{run_id}/request-revision/{approval_id}")
def request_revision(run_id: str, approval_id: str, body: ApprovalActionRequest = ApprovalActionRequest(), _: dict = Depends(require_admin)) -> dict[str, Any]:
    approval = _ensure_pending_approval_open(run_id, approval_id)
    with db_session() as db:
        db.execute(
            "UPDATE approval_requests SET status = 'revision_requested', resolved_at = CURRENT_TIMESTAMP, resolution_comment = ? WHERE id = ?",
            (body.note or None, approval_id),
        )
        if approval["workflow_step_run_id"]:
            db.execute(
                "UPDATE workflow_step_runs SET status = 'failed', completed_at = CURRENT_TIMESTAMP WHERE id = ?",
                (approval["workflow_step_run_id"],),
            )
            db.execute(
                "UPDATE agent_runs SET status = 'failed', completed_at = CURRENT_TIMESTAMP, error = ? WHERE id = ?",
                ("Approval returned for revision.", approval["workflow_step_run_id"]),
            )
        # mark the run as failed so the graph stops
        db.execute(
            "UPDATE workflow_runs SET status = 'failed', completed_at = CURRENT_TIMESTAMP WHERE id = ?",
            (run_id,),
        )
        db.execute(
            "INSERT INTO run_logs (id, workflow_run_id, level, message, metadata_json) VALUES (?, ?, 'warn', ?, ?)",
            (str(uuid4()), run_id, "Run stopped: approval returned for revision.", json.dumps({"approval_id": approval_id})),
        )
    return {"revision_requested": True, "approval_id": approval_id}


@router.post("/{run_id}/cancel")
def cancel_run(run_id: str, _: dict = Depends(require_admin)) -> dict[str, Any]:
    with db_session() as db:
        db.execute(
            "UPDATE workflow_runs SET status = 'cancelled', completed_at = CURRENT_TIMESTAMP WHERE id = ? AND status IN ('queued','running','waiting_approval')",
            (run_id,),
        )
    return {"cancelled": True, "run_id": run_id}


@router.get("/{run_id}/approvals")
def get_run_approvals(run_id: str, _: dict = Depends(require_user)) -> list[dict[str, Any]]:
    _expire_pending_approvals(run_id)
    with db_session() as db:
        rows = db.execute(
            "SELECT * FROM approval_requests WHERE workflow_run_id = ? ORDER BY created_at ASC",
            (run_id,),
        ).fetchall()
    return [
        {
            "id": r["id"],
            "status": r["status"],
            "title": r["title"],
            "reason": r["reason"],
            "context_summary": r["context_summary"],
            "workflow_step_run_id": r["workflow_step_run_id"],
            "created_at": r["created_at"],
            "expires_at": r["expires_at"],
            "resolved_at": r["resolved_at"],
        }
        for r in rows
    ]

File: internal/api/runs.go
// Handlers for the /api/workflow-runs READ surface, ported from
// backend/app/routers/runs.py.
//
// Starting a run and resolving an approval both resume execution through the
// engine and are not here — they need internal/exec wired in, and half a start
// endpoint is worse than none.
package api

import (
	"database/sql"
	"encoding/json"
	"errors"
	"fmt"
	"net/http"
	"strconv"

	"github.com/go-chi/chi/v5"
)

type run struct {
	ID            string          `json:"id"`
	WorkflowID    string          `json:"workflow_id"`
	Status        string          `json:"status"`
	TriggerType   *string         `json:"trigger_type"`
	WorkspacePath *string         `json:"workspace_path"`
	Graph         json.RawMessage `json:"graph"`
	FinalReport   *string         `json:"final_report"`
	CreatedAt     string          `json:"created_at"`
	CompletedAt   *string         `json:"completed_at"`
}

const runColumns = `id, workflow_id, status, trigger_type, workspace_path, graph_json, final_report, created_at, completed_at`

func scanRun(row interface{ Scan(...any) error }) (run, error) {
	var r run
	var trigger, workspace, report, completed sql.NullString
	var graphJSON sql.NullString
	if err := row.Scan(&r.ID, &r.WorkflowID, &r.Status, &trigger, &workspace,
		&graphJSON, &report, &r.CreatedAt, &completed); err != nil {
		return run{}, err
	}
	graph := graphJSON.String
	if graph == "" {
		graph = "{}"
	}
	r.Graph = json.RawMessage(graph)
	for _, pair := range []struct {
		src sql.NullString
		dst **string
	}{{trigger, &r.TriggerType}, {workspace, &r.WorkspacePath}, {report, &r.FinalReport}, {completed, &r.CompletedAt}} {
		if pair.src.Valid {
			value := pair.src.String
			*pair.dst = &value
		}
	}
	return r, nil
}

// clampInt keeps a caller-supplied bound inside a sane range rather than
// rejecting it — the UI passes these straight through.
func clampInt(raw string, fallback, min, max int) int {
	if raw == "" {
		return fallback
	}
	value, err := strconv.Atoi(raw)
	if err != nil {
		return fallback
	}
	if value < min {
		return min
	}
	if value > max {
		return max
	}
	return value
}

// listRuns returns recent runs, newest first.
//
// The limit is a caller-visible parameter rather than a hidden 20: callers were
// treating the truncated page as a total, so a dashboard tile froze at 20 once
// the 21st run existed. Anything wanting real totals uses /stats.
func (d *Deps) listRuns(w http.ResponseWriter, r *http.Request) {
	limit := clampInt(r.URL.Query().Get("limit"), 100, 1, 500)
	workflowID := r.URL.Query().Get("workflow_id")

	var rows *sql.Rows
	var err error
	if workflowID != "" {
		rows, err = d.Store.DB().Query(
			`SELECT `+runColumns+` FROM workflow_runs WHERE workflow_id = ? ORDER BY created_at DESC LIMIT ?`,
			workflowID, limit)
	} else {
		rows, err = d.Store.DB().Query(
			`SELECT `+runColumns+` FROM workflow_runs ORDER BY created_at DESC LIMIT ?`, limit)
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Could not list runs")
		return
	}
	defer rows.Close()

	out := []run{}
	for rows.Next() {
		item, err := scanRun(rows)
		if err != nil {
			writeError(w, http.StatusInternalServerError, "Could not read runs")
			return
		}
		out = append(out, item)
	}
	writeJSON(w, http.StatusOK, out)
}

func (d *Deps) getRun(w http.ResponseWriter, r *http.Request) {
	item, err := scanRun(d.Store.DB().QueryRow(
		`SELECT `+runColumns+` FROM workflow_runs WHERE id = ?`, chi.URLParam(r, "runID")))
	if errors.Is(err, sql.ErrNoRows) {
		writeError(w, http.StatusNotFound, "Run not found")
		return
	}
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Could not read the run")
		return
	}
	writeJSON(w, http.StatusOK, item)
}

type runStep struct {
	ID          string  `json:"id"`
	NodeID      string  `json:"node_id"`
	NodeType    string  `json:"node_type"`
	AgentName   *string `json:"agent_name"`
	AgentRole   *string `json:"agent_role"`
	Status      string  `json:"status"`
	Summary     *string `json:"summary"`
	Error       *string `json:"error"`
	StartedAt   *string `json:"started_at"`
	CompletedAt *string `json:"completed_at"`
}

// runSteps reads from agent_runs, NOT workflow_step_runs.
//
// The UI's notion of a "step" is the agent run: agent_name, agent_role, summary
// and error all live there. workflow_step_runs contributes only node_type, and
// the two share an id (ws.id = ar.id) rather than being related by a foreign
// key — which is why this is a LEFT JOIN on the primary key and not a lookup.
func (d *Deps) runSteps(w http.ResponseWriter, r *http.Request) {
	rows, err := d.Store.DB().Query(
		`SELECT ar.id, ar.node_id, COALESCE(ws.node_type, '') AS node_type,
		        ar.agent_name, ar.agent_role, ar.status, ar.summary, ar.error,
		        ar.started_at, ar.completed_at
		   FROM agent_runs ar
		   LEFT JOIN workflow_step_runs ws ON ws.id = ar.id
		  WHERE ar.workflow_run_id = ?
		  ORDER BY ar.started_at ASC`,
		chi.URLParam(r, "runID"))
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Could not read steps")
		return
	}
	defer rows.Close()

	out := []runStep{}
	for rows.Next() {
		var s runStep
		var agentName, agentRole, summary, errText, startedAt, completedAt sql.NullString
		if err := rows.Scan(&s.ID, &s.NodeID, &s.NodeType, &agentName, &agentRole,
			&s.Status, &summary, &errText, &startedAt, &completedAt); err != nil {
			writeError(w, http.StatusInternalServerError, "Could not read steps")
			return
		}
		for _, pair := range []struct {
			src sql.NullString
			dst **string
		}{{agentName, &s.AgentName}, {agentRole, &s.AgentRole}, {summary, &s.Summary},
			{errText, &s.Error}, {startedAt, &s.StartedAt}, {completedAt, &s.CompletedAt}} {
			if pair.src.Valid {
				value := pair.src.String
				*pair.dst = &value
			}
		}
		out = append(out, s)
	}
	writeJSON(w, http.StatusOK, out)
}

type runLog struct {
	ID        string          `json:"id"`
	Level     string          `json:"level"`
	Message   string          `json:"message"`
	Metadata  json.RawMessage `json:"metadata"`
	CreatedAt string          `json:"created_at"`
}

func (d *Deps) runLogs(w http.ResponseWriter, r *http.Request) {
	rows, err := d.Store.DB().Query(
		`SELECT id, level, message, metadata_json, created_at
		   FROM run_logs WHERE workflow_run_id = ? ORDER BY created_at`,
		chi.URLParam(r, "runID"))
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Could not read logs")
		return
	}
	defer rows.Close()

	out := []runLog{}
	for rows.Next() {
		var entry runLog
		var id, level sql.NullString
		var metadata sql.NullString
		if err := rows.Scan(&id, &level, &entry.Message, &metadata, &entry.CreatedAt); err != nil {
			writeError(w, http.StatusInternalServerError, "Could not read logs")
			return
		}
		entry.ID = id.String
		entry.Level = level.String
		// The column is metadata_json; the field is metadata, decoded.
		raw := metadata.String
		if raw == "" {
			raw = "{}"
		}
		entry.Metadata = json.RawMessage(raw)
		out = append(out, entry)
	}
	writeJSON(w, http.StatusOK, out)
}

type agentMessage struct {
	ID         string `json:"id"`
	AgentRunID string `json:"agent_run_id"`
	SenderType string `json:"sender_type"`
	SenderName string `json:"sender_name"`
	Content    string `json:"content"`
	CreatedAt  string `json:"created_at"`
}

// stepMessages returns the messages for one step. The step id IS the agent run
// id (see runSteps), so this filters on agent_run_id directly rather than
// joining back through the run.
func (d *Deps) stepMessages(w http.ResponseWriter, r *http.Request) {
	rows, err := d.Store.DB().Query(
		`SELECT id, agent_run_id, sender_type, sender_name, content, created_at
		   FROM agent_messages WHERE agent_run_id = ? ORDER BY created_at ASC`,
		chi.URLParam(r, "stepID"))
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Could not read messages")
		return
	}
	defer rows.Close()

	out := []agentMessage{}
	for rows.Next() {
		var m agentMessage
		if err := rows.Scan(&m.ID, &m.AgentRunID, &m.SenderType, &m.SenderName,
			&m.Content, &m.CreatedAt); err != nil {
			writeError(w, http.StatusInternalServerError, "Could not read messages")
			return
		}
		out = append(out, m)
	}
	writeJSON(w, http.StatusOK, out)
}

// runStats computes aggregates in SQL over the whole table rather than over a
// page of it. Counting from a truncated list is how the attention banner could
// report "All clear" while failures sat just outside the loaded window.
func (d *Deps) runStats(w http.ResponseWriter, r *http.Request) {
	windowHours := clampInt(r.URL.Query().Get("window_hours"), 24, 1, 24*30)
	since := fmt.Sprintf("-%d hours", windowHours)

	var total, failed, completed int
	err := d.Store.DB().QueryRow(
		`SELECT COUNT(*),
		        COALESCE(SUM(status = 'failed'), 0),
		        COALESCE(SUM(status = 'completed'), 0)
		   FROM workflow_runs WHERE created_at >= datetime('now', ?)`, since).
		Scan(&total, &failed, &completed)
	if err != nil {
		writeError(w, http.StatusInternalServerError, "Could not compute stats")
		return
	}

	// Active is counted with NO window: a run started two days ago and still
	// going is exactly what an operator needs to see.
	var active, waiting int
	d.Store.DB().QueryRow(
		`SELECT COUNT(*) FROM workflow_runs WHERE status IN ('running','queued','waiting_approval')`).Scan(&active)
	d.Store.DB().QueryRow(
		`SELECT COUNT(*) FROM workflow_runs WHERE status = 'waiting_approval'`).Scan(&waiting)

	// Oldest thing still running. "3 running" says nothing about whether one has
	// been stuck for an hour, which is the question an operator actually has.
	var oldest sql.NullString
	d.Store.DB().QueryRow(
		`SELECT MIN(created_at) FROM workflow_runs
		  WHERE status IN ('running','queued','waiting_approval')`).Scan(&oldest)

	currentMedian := d.medianDuration(since, "")
	// The same median over the PREVIOUS window, so the tile shows a direction
	// rather than a bare number.
	previousMedian := d.medianDuration(fmt.Sprintf("-%d hours", windowHours*2), since)

	body := map[string]any{
		"total": total, "failed": failed, "completed": completed,
		"active": active, "waiting_approval": waiting,
		"oldest_active_started_at":         nullableString(oldest),
		"median_duration_seconds":          currentMedian,
		"previous_median_duration_seconds": previousMedian,
		// Null unless BOTH windows have data. Guarding only the previous window
		// let an EMPTY current window report "81s faster" — a trend computed
		// against nothing, which is the same invention the guard exists to stop,
		// just from the other side.
		"median_delta_seconds": nil,
	}
	if currentMedian != nil && previousMedian != nil {
		delta := *currentMedian - *previousMedian
		body["median_delta_seconds"] = roundTo1(delta)
	}
	writeJSON(w, http.StatusOK, body)
}

// medianDuration is the true median, not the mean: one 40-minute outlier drags
// an average somewhere no actual run has ever been. The LIMIT/OFFSET pair is
// the standard SQLite median — LIMIT 2 minus the parity of the count picks one
// row for an odd count and two for an even one, and AVG then averages them.
func (d *Deps) medianDuration(since, until string) *float64 {
	where := `completed_at IS NOT NULL AND created_at >= datetime('now', ?)`
	args := []any{since}
	if until != "" {
		where += ` AND created_at < datetime('now', ?)`
		args = append(args, until)
	}

	query := `
		SELECT AVG(secs) FROM (
		  SELECT (julianday(completed_at) - julianday(created_at)) * 86400 AS secs
		    FROM workflow_runs WHERE ` + where + `
		   ORDER BY secs
		   LIMIT 2 - (SELECT COUNT(*) FROM workflow_runs WHERE ` + where + `) % 2
		  OFFSET (SELECT (COUNT(*) - 1) / 2 FROM workflow_runs WHERE ` + where + `)
		)`
	full := append(append(append([]any{}, args...), args...), args...)

	var median sql.NullFloat64
	if err := d.Store.DB().QueryRow(query, full...).Scan(&median); err != nil || !median.Valid {
		return nil
	}
	value := roundTo1(median.Float64)
	return &value
}

func roundTo1(value float64) float64 {
	return float64(int64(value*10+0.5)) / 10
}

func nullableString(value sql.NullString) any {
	if !value.Valid {
		return nil
	}
	return value.String
}
